### ISSUE TITLE
better handling of unmanageable trees?

### ISSUE THREAD

**Issue by adf-ncgr** [legumeinfo/Funnotate#27]
Some trees are forests unto themselves, and the user will wait a long time to get any satisfaction if they are lucky with something like: https://funnotate.legumeinfo.org/?family=legfed_v1_0.L_009642

it is probably worth noting that although the tree is large, it actually comes back reasonably quickly in the older app:
https://legacy.legumeinfo.org/chado_phylotree/legfed_v1_0.L_009642

Not sure if the excessive slowness in the new version is a consequence of using lorax or something else. But, at a minimum it might be nice if the gene family descriptor could be displayed before the whole tree is obtained, so the patient user can decide based on that descriptor if the wait is likely to be worth their while.

**Comment by svengato** [comment]
With the message "Are you sure you want Birnam Forest to come to Funnotate?"

**Comment by adf-ncgr** [comment]
All Hail MacVyvyan!

**Comment by adf-ncgr** [comment]
It's like adding injury to insult when the page that you waited for so long to load that you started ignoring it and doing other things upon your going back to check on it is giving you the shiny salute: 
```
Disconnected from the server.
Reload
```

**Comment by svengato** [comment]
Retrieving the tree takes less than a second, displaying the tree is the bottleneck. I am working on speeding it up.

**Comment by adf-ncgr** [comment]
OK, that's good info- I thought the older app and the funnotate version were basically using the same display engine, so it seems a little weird that this would be the bottleneck, but I'm probably oversimplifying the way things evolved.

**Comment by svengato** [comment]
It looks like TnT Tree's parse_newick() (which converts the Newick tree to a JSON object) is the problem. I am looking into that, maybe we will have to write our own (faster/more robust) version.

**Comment by adf-ncgr** [comment]
I guess that makes sense. The older version was doing conversion to json server-side, and I think it was using a pre-parsed database representation of the tree structure as well. However, it does seem like newick parsing ought not to be a major bottleneck, though perhaps nested parentheses are more of a syntactical viper's nest than I'm imagining. Anyway, sounds like you're on a reasonable course. It might also be possible to get the data in a non-newick form from intermine (possibly not the json you'd want, however), but being able to handle a standard format seems better.

**Comment by adf-ncgr** [comment]
I suppose we could avoid optimizing the parser if we instead stored pre-JSONified versions of the trees and served them as an option from lorax. Not sure if that would end up being an easier route, but it would be more similar in spirit to how the older app worked. Just a thought, not sure on which side of the fine line it may lie between clever and stupid

**Comment by svengato** [comment]
On closer examination, the parser is fast but loading the parsed data into the tree for visualization is the bottleneck. (My mistake, they were on the same line.) I am looking for a way around this.

**Comment by adf-ncgr** [comment]
does it differ in some essential way from how the older app did the corresponding thing? the only thing that comes to mind as a possibility (based on my limited familiarity with both old and new apps) would be if perhaps some event handling code  is getting prematurely associated and going through a lot of work firing / receiving before the tree is actually ready for interaction.

**Comment by svengato** [comment]
There is a line
```
// wrap in timeout in order to wait to render the tree before computing label bounding boxes
setTimeout(highlightSequences, TIMEOUT_DURATION_MSEC);
```
but this is for highlighting user sequences, it should not apply to the family-based tree.

I will look at what the legacy code did.

**Comment by svengato** [comment]
Try it now -

**Comment by svengato** [comment]
Bitte schnell, nein? But subsequent toggles of the taxa chart and MSA no longer work, it looks like the global copy of the phylotree is not updating properly. I am looking at that now.

**Comment by adf-ncgr** [comment]
Ja, schnell, schnell- aber wo sind deine komplimente?

**Comment by adf-ncgr** [comment]
what did you change to make the improvement?

**Comment by svengato** [comment]
I just wrapped the tree initialization in a future block (in R). But I think the promise handling really needs to go in the JavaScript itself. No kompliments until then!

**Comment by svengato** [comment]
Still struggling with this. It needs something like Crossfilter, where we set all the data first and then try to draw the charts.

**Comment by adf-ncgr** [comment]
IIRC, the original phylotree implementation used dc.js which I think uses crossfilter somehow. In any case, maybe best to put production version back to the sometimes slow but always fully functional state? I was actually wanting to look at one of the MSAs today and couldn't (that functionality seems to be completely broken on the legacy site now, although it may be a simple fix there, looks like an error similar to what you helped Ethy get around).

**Comment by svengato** [comment]
Production version is back where it was.

**Comment by adf-ncgr** [comment]
Thanks, that gave me what I needed.

**Comment by svengato** [comment]
"The asynchronous processes win, they always do."

**Comment by svengato** [comment]
Finally figured out how to set the 'hidden' attribute for all MSA sequences simultaneously instead of individually. Fixed in commit [0249277](https://github.com/legumeinfo/Funnotate/commit/0249277e870bb6386c0c16cce3f10afd157df5f4).
